File: sdk/python/src/openagentio/adapter/http/options.py
"""Adapter options mirroring pkg/adapter/http/options.go."""
from __future__ import annotations

import logging
from dataclasses import dataclass, field
from typing import Any, Callable, TYPE_CHECKING

if TYPE_CHECKING:
    from openagentio.adapter.http.auth import AuthFunc
    from openagentio.adapter.http.middleware import ASGIMiddleware


@dataclass
class AdapterOptions:
    auth: AuthFunc | None = None
    logger: logging.Logger | None = None
    timeout: float = 30.0
    idle_timeout: float = 0.0
    middleware: list[ASGIMiddleware] = field(default_factory=list)


Option = Callable[[AdapterOptions], None]


def WithAuth(fn: AuthFunc) -> Option:
    def apply(o: AdapterOptions) -> None:
        o.auth = fn
    return apply


def WithLogger(l: logging.Logger) -> Option:
    def apply(o: AdapterOptions) -> None:
        if l is not None:
            o.logger = l
    return apply


def WithTimeout(d: float) -> Option:
    def apply(o: AdapterOptions) -> None:
        o.timeout = d
    return apply


def WithIdleTimeout(d: float) -> Option:
    def apply(o: AdapterOptions) -> None:
        o.idle_timeout = d
    return apply


def WithMiddleware(*mws: ASGIMiddleware) -> Option:
    def apply(o: AdapterOptions) -> None:
        o.middleware.extend(mws)
    return apply

File: pkg/adapter/http/options.go
package http

import (
	"log/slog"
	"time"
)

// Option mutates an Adapter at construction time.
type Option func(*Adapter)

// WithAuth installs a request authenticator. When set, every request must
// pass the AuthFunc; non-nil error short-circuits with 401. A nil AuthFunc
// (or never calling WithAuth) leaves the adapter unauthenticated.
func WithAuth(fn AuthFunc) Option {
	return func(a *Adapter) { a.auth = fn }
}

// WithLogger overrides the default slog.Default() logger.
func WithLogger(l *slog.Logger) Option {
	return func(a *Adapter) {
		if l != nil {
			a.log = l
		}
	}
}

// WithTimeout caps the per-request overall deadline (invoke + stream).
// Passes through to bus.WithTimeout. Zero leaves the bus default in effect.
func WithTimeout(d time.Duration) Option {
	return func(a *Adapter) { a.timeout = d }
}

// WithIdleTimeout caps the gap between two streaming frames. Zero disables
// idle-timeout, leaving only the overall WithTimeout in effect.
func WithIdleTimeout(d time.Duration) Option {
	return func(a *Adapter) { a.idle = d }
}

// WithMiddleware appends one or more http.Handler middlewares. Outermost
// middleware (first in the list) wraps the others.
func WithMiddleware(mw ...Middleware) Option {
	return func(a *Adapter) { a.mws = append(a.mws, mw...) }
}

// WithSSERetry sets the retry interval sent to SSE clients in the retry:
// field (milliseconds). Zero disables the field. Default is 3s.
func WithSSERetry(d time.Duration) Option {
	return func(a *Adapter) { a.sseRetry = d }
}
